### PR TITLE
Implement draining logic for transactional flow

### PR DESCRIPTION
**Pull request by szymonm**
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](https://github.com/akka/alpakka-kafka/tree/master/CONTRIBUTING.md)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [ ] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?

## Fixes

Duplications in a transactional flow that occur, when either closing the stream or partitions are revoked by Kafka.

## Purpose

This PR introduces draining logic for `TransactionalSource`. Specifically, for each partition managed by the source, we keep track of the message offsets emitted by the source.
When closing the stage or revoking partitions, we make sure that all the offsets are committed back to the Kafka by the producer.
To know that an offset was committed by the producer we attach `CommittedMarker` to each message produced by the stream. `TransactionalProducerStage` uses this marker to tell the source that an offset was committed.

## Background Context

In a transactional stream processing partition `P`, we need to wait for all commits to be acknowledged by Kafka before closing the consumer.
If we don't, it's possible that another consumer is assigned `P`, which triggers fetching offset that can be stale or overwritten later. This causes data duplication.

## References

https://github.com/akka/alpakka-kafka/issues/758
https://github.com/akka/alpakka-kafka/issues/756

### CODE REVIEW

**Comment by szymonm** [comment]
Please don't merge this PR. It's only to demonstrate the issue.

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 1]
```diff
@@ -146,8 +163,10 @@ private final class TransactionalProducerStageLogic[K, V, P](stage: Transactiona
 
   override val onMessageAckCb: AsyncCallback[Envelope[K, V, P]] =
     getAsyncCallback[Envelope[K, V, P]](_.passThrough match {
-      case o: ConsumerMessage.PartitionOffset => batchOffsets = batchOffsets.updated(o)
-      case _ =>
+      case o: ConsumerMessage.PartitionOffsetCommittedMarker =>
+        batchOffsets = batchOffsets.updated(o)
+        println(s"ADDING OFF ${o.offset}")
+      case x => println("Chuj " + x)
```
Comment (2m) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: I think your frustration might be caused by another bug with offset tracking. With a fix for that applied, your test case does not fail for me anymore. The fix is this one: https://github.com/akka/alpakka-kafka/pull/752/commits/d381b4f451e76f125a39424d62ea243eebbae0f3

Comment (szymonm) [reply core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: Sorry, this should not be included in the PR. Will rewrite history.

**Comment by szymonm** [comment]
Some afterthoughts:
IIUC, we need to ensure that consumer is not closed before producer finishes committing transaction also in the case of stream failure.
IIUC, in this case the failure message will reach source stage and stop the consumer. However, it will not reach `TransactionalProducerStage` if `producer.commitTransaction` will be in progress (it's blocking). This can lead to race conditions, when producer commits after some other consumer is assigned the partition.

File: core/src/main/scala/akka/kafka/internal/TransactionalSource.scala [hunk 2]
```diff
@@ -36,9 +81,73 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
     IsolationLevel.READ_COMMITTED.toString.toLowerCase(Locale.ENGLISH)
   )
 
+  case object Drained
+  case class Drain[T](ack: ActorRef, msg: T)
+  case class Committed(offsets: Iterable[ConsumerMessage.PartitionOffset])
```
Comment (2m) [inline core/src/main/scala/akka/kafka/internal/TransactionalSource.scala]: I do not see that this one would be handled in the stage actor

**Comment by 2m** [comment]
Rebased on top of the latest master, and also fixed a bug, where a wrong `Committed` message was being handled in the Trnsational Source stage.

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 3]
```diff
@@ -151,9 +151,9 @@ private final class TransactionalProducerStageLogic[K, V, P](stage: Transactiona
   private def maybeCommitTransaction(beginNewTransaction: Boolean = true): Unit = {
     val awaitingConf = awaitingConfirmation.get
     batchOffsets match {
-      case batch: NonemptyTransactionBatch if awaitingConf == 0 =>
+      case batch: NonemptyTransactionBatch if awaitingConf == 0 && !hasBeenPulled(stage.in) =>
```
Comment (szymonm) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: Can you explain why we need this logic?

**Comment by szymonm** [comment]
Thanks, @2m

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala (hunk 3 above)
Comment (2m) [reply core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: With this PR, the transaction commit is split into two synchronous blocks in the Producer stage. That means that in between, a message could be sent to the stage which would be sent to the Kafka producer, but the transaction has not yet been started.

So this guard was needed to be sure that no message will come in between the the committing of transaction and beginning of the new one.

File: core/src/main/scala/akka/kafka/internal/TransactionalSource.scala [hunk 4]
```diff
@@ -36,9 +81,73 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
...
+      override def partitionRevokedCB(revokedTps: Set[TopicPartition]): Unit = {
+        waitForDraining()
+        sourceActor.ref ! Revoked(revokedTps.toList)
+        super.partitionRevokedCB(revokedTps)
+      }
+
+      def waitForDraining(): Unit = {
+        import akka.pattern.ask
+        implicit val timeout = Timeout(5.seconds)
+        Await.result(ask(stageActor.ref, Drain(stageActor.ref, Drained)), 10.seconds)
```
Comment (jotek7) [inline core/src/main/scala/akka/kafka/internal/TransactionalSource.scala]: I'm afraid if the behaviour in case of a timeout is correct.
IIUC TimeoutException is thrown and it's caught in kafka code by ConsumerCoordinator,
here:

```
try {
            listener.onPartitionsRevoked(revoked);
        } catch (WakeupException | InterruptException e) {
            throw e;
        } catch (Exception e) {
            log.error("User provided listener {} failed on partition revocation", listener.getClass().getName(), e);
        }
```

So due to `catch (Exception e)` we won't fail the stream and we'll just continue like nothing happened and the stream was drained.

Therefore if it takes too long to drain then we still allow messages to remain in the stream. In the meantime some other consumer might start processing data from the same partition because we just released it.

@szymonm @2m Could you verify if my reasoning is correct?

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala (hunk 3 above)
Comment (szymonm) [reply core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: Still didn't quite get it. I was hoping that suspending demand when transaction is in progress will be enough.
Also, I'm wondering if we just shouldn't block waiting for internal commit.
`kafkaProducer.commitTrasnaction()` is already blocking the whole stage waiting on a network call. And `internalCommit` is just waiting for local memory change, so it shouldn't add much...
That would simplify the code a bit.

Comment (szymonm) [reply core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: Making `internalCommit`just send a message would also be possible. Seems like `TransactionalProducerStage` does not have to wait for the SourceActor to acknowledge getting the message, right?

File: core/src/main/scala/akka/kafka/internal/TransactionalSource.scala (hunk 4 above)
Comment (szymonm) [reply core/src/main/scala/akka/kafka/internal/TransactionalSource.scala]: You're right @jotek7. In this case we should fail the stream starting from the producer. Which is similar to what we should do generally in the failure case (see my comments)

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala (hunk 3 above)
Comment (2m) [reply core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: > I was hoping that suspending demand when transaction is in progress will be enough.

Calling `suspendDemand` does make sure that `stage.in` is not going to be pulled anymore from that point on, but it could have been pulled from before.

> Seems like TransactionalProducerStage does not have to wait for the SourceActor to acknowledge getting the message, right?

Yes, currently it looks like that's the case. I'll need to think about it a bit more.

File: core/src/main/scala/akka/kafka/internal/TransactionalSource.scala [hunk 5]
```diff
@@ -36,9 +81,73 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
...
+      override def shuttingDownReceive = super.shuttingDownReceive.orElse(drainHandling)
+
+      def drainHandling: PartialFunction[(ActorRef, Any), Unit] = {
+        case (sender, Committed(offsets)) =>
+          inFlightRecords.committed(offsets.mapValues(_.offset() - 1))
+          sender ! Done
+        case (_, Drain(ack, msg)) =>
+          if (inFlightRecords.empty()) {
+            log.debug("Source drained")
+            ack ! msg
```
Comment (2m) [inline core/src/main/scala/akka/kafka/internal/TransactionalSource.scala]: This should also send an answer back to the `sender` otherwise the `ask` on line 139 will not be completed.

**Comment by 2m** [comment]
I added a commit where it fails the consumer stage if the draining timeout happens.

File: tests/src/test/scala/akka/kafka/scaladsl/TransactionsSpec.scala [hunk 6]
```diff
@@ -266,7 +266,7 @@ class TransactionsSpec extends SpecBase(kafkaPort = KafkaPorts.TransactionsSpec)
     "provide consistency when multiple transactional streams are being restarted" in {
       val sourcePartitions = 10
       val destinationPartitions = 4
-      val consumers = 3
+      val consumers = 1
```
Comment (szymonm) [inline tests/src/test/scala/akka/kafka/scaladsl/TransactionsSpec.scala]: why did you need to change this here?

**Comment by szymonm** [comment]
> I am not sure if this is a problem. If the stream is being torn down, then after the blocking `producer.commitTransaction` is done, it will complete the stage with `onCompletionFailure` and will not commit the next transaction, which might have duplicates. Am I missing something here?
There is still a chance that consumer will be closed before the producer finishes `producer.commitTransaction`. This can cause the partition to be picked up by another consumer that will start from the offsets that are no yet committed.

File: tests/src/test/scala/akka/kafka/scaladsl/TransactionsSpec.scala (hunk 6 above)
Comment (2m) [reply tests/src/test/scala/akka/kafka/scaladsl/TransactionsSpec.scala]: I noticed that this test-case does not finish when running with 3 consumers. The stopping of the stage on drain timeout does not work nicely with how this test uses the RestartSource and it gets restarted all the time and the test never finishes. I also noticed that in Travis it did not finish even with 1 consumer. I need to improve that test-case.

**Comment by szymonm** [comment]
What are the requirements for binary compatibility? After turning `PartitionOffset` case class to a normal class we have some warnings. Should we fix them all?

**Comment by 2m** [comment]
> Should we fix them all?

Yes, otherwise we will not be able to include this in a patch release.

**Comment by szymonm** [comment]
Last travis run reports: `The job exceeded the maximum time limit for jobs, and has been terminated.`

The tests for `TransacionsSpec` take a lot of time: 
```[info] - must provide consistency when multiple transactional streams are being restarted (8 minutes, 18 seconds)
[info] - must drain stream on partitions rebalancing (2 minutes, 10 seconds)```

**Comment by szymonm** [comment]
Looks like the test `provide consistency when multiple transactional streams are being restarted` takes 8 mins. We run it twice (for each scala version) and thus we increase the test times from ~30mins to over 50mins (the test for draining takes another 2 mins for each run).

Can we make the test case smaller, @2m ?

**Comment by 2m** [comment]
Yes, lets decrease the number of messages to 100k and see how long does that take.

File: core/src/main/scala/akka/kafka/internal/TransactionalSource.scala [hunk 7]
```diff
@@ -36,9 +82,85 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
...
+          sourceActor.ref ! Revoked(revokedTps.toList)
+        } else {
+          sourceActor.ref ! Failure(new Error("Timeout while draining"))
+        }
+        super.partitionRevokedCB(revokedTps)
+      }
+
+      def waitForDraining(): Boolean = {
+        import akka.pattern.ask
+        // Shall we use commitTimeout here?
```
Comment (2m) [inline core/src/main/scala/akka/kafka/internal/TransactionalSource.scala]: Yes, it should be at least commit timeout. Possible a bit larger as well. Can you refactor this from hardcoded constant to configuration value?

File: core/src/main/scala/akka/kafka/internal/TransactionalSource.scala [hunk 8]
```diff
@@ -36,9 +82,85 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
...
+          sender ! Done
+        case (sender, Drain(ack, msg)) =>
+          if (inFlightRecords.empty()) {
+            log.debug("Source drained")
+            ack.getOrElse(sender) ! msg
+          } else {
+            log.debug(s"Draining partitions {}", inFlightRecords)
+            // TODO: This timeout should be somehow related to the committing interval.
+            // For instance having 10.millis makes no sense if we commit transaction every second
+            materializer.scheduleOnce(30.millis, new Runnable {
```
Comment (2m) [inline core/src/main/scala/akka/kafka/internal/TransactionalSource.scala]: Would be good to move this to configuration as well.

File: core/src/main/scala/akka/kafka/internal/TransactionalSource.scala (hunk 7 above)
Comment (szymonm) [reply core/src/main/scala/akka/kafka/internal/TransactionalSource.scala]: After some thoughts I think using commitTimeout is just ok, will use that.

**Comment by 2m** [comment]
Two of the other tests in the TransactionalSpec started failing. Could be the new timing configuration not playing nicely with the test-cases.

**Comment by szymonm** [comment]
This seems to be a problem with `expectN` timeout of 10s by default. Is there a pattern how we could change it on a per-test basis, @2m?

**Comment by szymonm** [comment]
Alternatively, we can change it for all tests (not nice)...

**Comment by szymonm** [comment]
These tests actually catch a valid issue now:
We keep track of offsets emitted by the source and now we are waiting for all emitted offsets. But we should not wait for the offset emitted if processing of this element caused an exception.

**Comment by szymonm** [comment]
If produced is closed after committing a transaction we can safely close the consumer.

**Comment by szymonm** [comment]
Looks like `IntegrationSpec` `signal rebalance events to actor` test is flakky.

**Comment by szymonm** [comment]
TBH there is no reason why this test should work reliably. 
It uses `ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest"` config with no offsets storing.

This means that after a rebalancing consumers pick up the latest element from the queue. If there is more elements written to a ktp during the rebalancing this elements are never read.

**Comment by szymonm** [comment]
Can you run tests again?

**Comment by 2m** [comment]
> We keep track of offsets emitted by the source and now we are waiting for all emitted offsets. But we should not wait for the offset emitted if processing of this element caused an exception.

Great catch. That was the reason, I saw the multiple "waiting for draining" messages in the logs when the stage was stopping because of failure.

> Looks like IntegrationSpec signal rebalance events to actor test is flaky.

I do not see it failing anymore, but if we see that again, lets create an issue and fix it in a separate PR.

**Comment by szymonm** [comment]
> I do not see it failing anymore, but if we see that again, lets create an issue and fix it in a separate PR.
👍

**Comment by szymonm** [comment]
What are the next steps with the PR?

**Comment by 2m** [comment]
I am currently running the test with `consumers = 3` and after a little tweak, where we let the copy flow to restart a couple more times, it is passing. I am going to push that and if that looks green on Travis, this will be good to go in.

**Comment by 2m** [comment]
With multiple consumers the test fails with missing messages.

File: core/src/main/resources/reference.conf [hunk 9]
```diff
@@ -93,6 +93,10 @@ akka.kafka.consumer {
   # This value is used instead of Kafka's default from `default.api.timeout.ms`
   # which is 1 minute.
   metadata-request-timeout = 5s
+
+  # Interval for cheking that transaction was completed before closing the consumer.
```
Comment (seglo) [inline core/src/main/resources/reference.conf]: I don't see any more documentation for the end user about his new functionality.  I suggest updating the Transactions section in the docs.

Comment (seglo) [reply core/src/main/resources/reference.conf]: [Minor] Typo on "checking".

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 10]
```diff
@@ -159,21 +190,35 @@ private final class TransactionalProducerStageLogic[K, V, P](stage: Transactiona
...
-      scheduleOnce(commitSchedulerKey, commitInterval)
-    }
+    batch
+      .internalCommit()
+      .onComplete { _ =>
+        onInternallCommitAckCb.invoke(beginNewTransaction)
+      }(materializer.executionContext)
+  }
+
+  val onInternallCommitAckCb: AsyncCallback[Boolean] = {
```
Comment (seglo) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: [Minor] Typo `onInternalCommitAckCb`

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 11]
```diff
@@ -159,21 +190,35 @@ private final class TransactionalProducerStageLogic[K, V, P](stage: Transactiona
   override def onCompletionFailure(ex: Throwable): Unit = {
     log.debug("Aborting transaction due to stage failure")
     abortTransaction()
+    batchOffsets.committingFailed()
     super.onCompletionFailure(ex)
   }
 
   private def commitTransaction(batch: NonemptyTransactionBatch, beginNewTransaction: Boolean): Unit = {
     val group = batch.group
     log.debug("Committing transaction for consumer group '{}' with offsets: {}", group, batch.offsetMap())
```
Comment (seglo) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: [Minor] Use the local variable offsetMap so we don't have to to build for each log line?

File: core/src/main/scala/akka/kafka/internal/TransactionalSource.scala [hunk 12]
```diff
@@ -6,24 +6,73 @@
...
+                                                     commitTimeout: FiniteDuration)
     extends KafkaSourceStage[K, V, TransactionalMessage[K, V]](
       s"TransactionalSource ${subscription.renderStageAttribute}"
     ) {
   require(consumerSettings.properties(ConsumerConfig.GROUP_ID_CONFIG).nonEmpty, "You must define a Consumer group.id.")
 
+  type Offset = Long
+
+  private trait InFlightRecords {
+    // Assumes that offsets per topic partition are added in the increasing order
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalSource.scala]: What makes this assumption hold? Can you explain it in the comment, please.

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 13]
```diff
@@ -111,32 +132,44 @@ private final class TransactionalProducerStageLogic[K, V, P](stage: Transactiona
     setHandler(stage.out, new OutHandler {
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: These `setHandler` and `demandSuspended = false` should possibly become a method to stress they belong together.

File: core/src/main/scala/akka/kafka/internal/MessageBuilder.scala [hunk 14]
```diff
@@ -96,6 +101,16 @@ private[kafka] trait InternalCommitter {
   def commit(batch: CommittableOffsetBatch): Future[Done]
 }
 
+/** Internal API */
+@InternalApi
+private[kafka] trait CommittedMarker {
+  // Marks offsets as already committed
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/MessageBuilder.scala]: These should be scaladoc-style comments.

File: core/src/main/scala/akka/kafka/ConsumerSettings.scala [hunk 15]
```diff
@@ -540,5 +554,6 @@ class ConsumerSettings[K, V] @InternalApi private[kafka] (
     s"commitTimeWarning=${commitTimeWarning.toCoarsest}," +
     s"waitClosePartition=${waitClosePartition.toCoarsest}," +
     s"metadataRequestTimeout=${metadataRequestTimeout.toCoarsest}" +
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/ConsumerSettings.scala]: Comma missing at the end.

File: core/src/main/resources/reference.conf [hunk 16]
```diff
@@ -93,6 +93,10 @@ akka.kafka.consumer {
   # This value is used instead of Kafka's default from `default.api.timeout.ms`
   # which is 1 minute.
   metadata-request-timeout = 5s
+
+  # Interval for checking that transaction was completed before closing the consumer.
+  # Used in the transactional flow.
+  draining-check-interval = 30ms
```
Comment (ennru) [inline core/src/main/resources/reference.conf]: I wonder if the setting name should contain `eos` or something. There are so many settings now it is hard to see which belong together.

File: core/src/main/scala/akka/kafka/ConsumerMessage.scala [hunk 17]
```diff
@@ -78,16 +79,53 @@ object ConsumerMessage {
...
   }
 
   /**
    * Offset position and metadata for a groupId, topic, partition.
    */
   final case class PartitionOffsetMetadata(key: GroupTopicPartition, offset: Long, metadata: String)
 
+  final case class PartitionOffsetCommittedMarker(override val key: GroupTopicPartition,
+                                                  override val offset: Long,
+                                                  private[kafka] val committer: CommittedMarker)
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/ConsumerMessage.scala]: I'd prefer `committedMarker` as name to distinguish it from `Committer`.

File: tests/src/test/scala/akka/kafka/scaladsl/TransactionsSpec.scala [hunk 18]
```diff
@@ -349,6 +356,120 @@ class TransactionsSpec extends SpecBase(kafkaPort = KafkaPorts.TransactionsSpec)
...
+      val probeConsumerGroup = createGroupId(2)
+      val probeConsumer = valuesProbeConsumer(probeConsumerSettings(probeConsumerGroup), sinkTopic)
+
+      periodicalCheck("Wait for elements written to Kafka", maxTries = 30, 1.second) {
+        elementsWrote.get
+      }(_ > 10)
+
+      probeConsumer
+        .request(elements)
+        .expectNextUnorderedN((1 to elements).map(_.toString))
```
[the remark this answers is not in the document]
Comment (szymonm) [reply tests/src/test/scala/akka/kafka/scaladsl/TransactionsSpec.scala]: We are reading from multiple partitions, kafka gives us a guarantee that messages in a single partition are ordered.

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala (hunk 13 above)
Comment (szymonm) [reply core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala]: We have methods `resumeDemand` and `suspendDemand` now. Is this what you were thinking of?

File: core/src/main/scala/akka/kafka/internal/TransactionalSource.scala (hunk 12 above)
Comment (szymonm) [reply core/src/main/scala/akka/kafka/internal/TransactionalSource.scala]: Added:
```
    // The assumption is true for a Kafka consumer that guarantees that elements are emitted
    // per partition in offset-increasing order.```

File: core/src/main/resources/reference.conf (hunk 16 above)
Comment (szymonm) [reply core/src/main/resources/reference.conf]: tbh... `eos` is cryptic even for me...

Comment (szymonm) [reply core/src/main/resources/reference.conf]: but we may have some distinct section for transactional processing?

**Comment by szymonm** [comment]
The test for the failure case is still failing showing both duplication and data loss.

My intuition what happens is as follows:
1. Data loss
Since we send messages and offsets in transaction the only possibility to loose data is to loose some records while they are in the stream and commit records with higher offsets.

2. Duplication
We close consumer before the producer in a single stream. Thus another consumer starts consuming from the partition before the previous producer flushes last offsets.

**Comment by szymonm** [comment]
I'm also seeing some problems with the draining mechanism, specifically it happens that it waits fvor the last message that is never drained.

**Comment by szymonm** [comment]
Also setting higher backoff for restarts increases the probability of test success (from what I observed)
```RestartSource
          .onFailuresWithBackoff(1.second, 5.seconds, 0.2)```

**Comment by 2m** [comment]
> Since we send messages and offsets in transaction the only possibility to loose data is to loose some records while they are in the stream and commit records with higher offsets.

I only noticed lost messages when running that test-case with multiple consumers. Investigating that right now.

> Duplication

I have not seen test-case failing because of duplication for quite some time now. The duplication check is first, and at least in Travis it passes now for quite some time before the test-case fails with missing elements check.

**Comment by 2m** [comment]
Pushed two more commits that improve the failing test-case in a couple of ways. When it comes to duplicates, it could very well be, that we get duplicated messages when rebalance happens while consuming messages after transactional processing. For that I added offset tracking which will ignore duplicates that have same offsets. We should only care about duplicates with different offsets, because those would be from the transactions.

Also I moved the place where the transactional processing flow is terminated in the test stream, so the stream is always terminated with an error (from upstream) instead of cancellation (from downstream). Lets see what Travis thinks about that when it comes to missing messages.

**Comment by szymonm** [comment]
I'm looking into it too, and what worries me know is seeing `java.lang.Error: Timeout while draining`. My intuition is that with embedded kafka and timeout set to 15s we have plenty of time to drain.
Maybe timeout indicates that we don't pull correctly while draining?

**Comment by szymonm** [comment]
Also, when the timeout happens, we should make sure that close the producer before closing the consumer. This doesn't happen, right?

**Comment by 2m** [comment]
> Also, when the timeout happens, we should make sure that close the producer before closing the consumer.

Not explicitly, but there is a `akka.kafka.consumer.stop-timeout` that delays stopping the ConsumerStage so late commits can happen when using a CommittableSource.

I see that we are overriding `stopConsumerActor` in TransactionalSource and not doing the delay anymore. We should probably bring the delay back so the producer gets closed first.

> Maybe timeout indicates that we don't pull correctly while draining?

Yea, noticed those timeouts as well. But I also see `Source drained.` in the logs. Could it be that the reply after draining does not come back to complete the ask Future?

**Comment by szymonm** [comment]
> Yea, noticed those timeouts as well. But I also see Source drained. in the logs. Could it be that the reply after draining does not come back to complete the ask Future?

That was another idea I had!

**Comment by szymonm** [comment]
After increasing the parallelism of sending messages to kafka we don't have to wait for messages to be flushed. As a result we are not failing stream draining that often, which gives very good chance of the test to pass (didn't see a single failure on my machine after changing that).

It also speeds up the test a lot!

**Comment by 2m** [comment]
Looks good. I also like the simplification in the Producer where the transaction is started sooner. That simplification means that these two commits are not needed anymore, because transaction is committed and new transaction is started in the same pass.

https://github.com/akka/alpakka-kafka/pull/757/commits/7be3450b28f140d75a0d0ff626b9a64f66178945
https://github.com/akka/alpakka-kafka/pull/757/commits/9c04dcf2a8d730b54ac9e9cd35b025ff4a544e6c

**Comment by 2m** [comment]
Pushed a revert of those two commits. I'll do a docs update and then this will be good to go.

**Comment by szymonm** [comment]
Yeap, happy to see this simplified again!

File: core/src/main/scala/akka/kafka/ConsumerMessage.scala [hunk 19]
```diff
@@ -78,16 +79,53 @@ object ConsumerMessage {
...
+
+    def unapply(arg: PartitionOffset): Option[(GroupTopicPartition, Long)] = Some((arg.key, arg.offset))
   }
 
   /**
    * Offset position and metadata for a groupId, topic, partition.
    */
   final case class PartitionOffsetMetadata(key: GroupTopicPartition, offset: Long, metadata: String)
 
+  final case class PartitionOffsetCommittedMarker(override val key: GroupTopicPartition,
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/ConsumerMessage.scala]: Internal API.

File: core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala [hunk 20]
```diff
@@ -78,15 +71,30 @@ import scala.concurrent.{Future, Promise}
...
+      completeStage()
+  }
+
+  protected def stopConsumerActor(): Unit =
     materializer.scheduleOnce(settings.stopTimeout, new Runnable {
       override def run(): Unit =
         consumerActor.tell(KafkaConsumerActor.Internal.Stop, sourceActor.ref)
     })
+
+  protected def partitionAssignedCB(assignedTps: Set[TopicPartition]): Unit = {
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala]: I find the `CB` strange as this is not an `AsyncCallback`, maybe `Signal` or `Handler`?

File: docs/src/main/paradox/transactions.md [hunk 21]
```diff
@@ -30,12 +30,14 @@ Transactions are committed on an interval which can be controlled with the produ
 
 When the stream is materialized the producer will initialize the transaction for the provided `transactional.id` and a transaction will begin.  Every commit interval (`eos-commit-interval`) we check if there are any offsets available to commit.  If offsets exist then we suspend backpressured demand while we drain all outstanding messages that have not yet been successfully acknowledged (if any) and then commit the transaction.  After the commit succeeds a new transaction is begun and we re-initialize demand for upstream messages.
 
+Messages are also drained from the stream when the consumer gets a rebalance of partitions. In that case, the consumer will wait in the `onPartitionsRevoked` callback until all of the messages have been drained from the stream and transaction is committed before allowing the rebalance to continue. The amount of total time the consumer will wait for draining is controlled by the `akka.kafka.consumer.commit-timeout`, and the interval between checks is controlled by the `akka.kafka.consuner.eos-draining-check-interval` configuration settings.
```
Comment (ennru) [inline docs/src/main/paradox/transactions.md]: "the" transaction

```suggestion
Messages are also drained from the stream when the consumer gets a rebalance of partitions. In that case, the consumer will wait in the `onPartitionsRevoked` callback until all of the messages have been drained from the stream and the transaction is committed before allowing the rebalance to continue. The amount of total time the consumer will wait for draining is controlled by the `akka.kafka.consumer.commit-timeout`, and the interval between checks is controlled by the `akka.kafka.consuner.eos-draining-check-interval` configuration settings.
```

File: docs/src/main/paradox/transactions.md [hunk 22]
```diff
@@ -30,12 +30,14 @@ Transactions are committed on an interval which can be controlled with the produ
...
+
 To gracefully shutdown the stream and commit the current transaction you must call `shutdown()` on the `Control` (@scala[@scaladoc[API](akka.kafka.scaladsl.Consumer$$Control)]@java[@scaladoc[API](akka.kafka.javadsl.Consumer$$Control)]) materialized value to await all produced message acknowledgements and commit the final transaction.  
 
 ### Simple Example
 
 Scala
-: @@ snip [snip](/tests/src/test/scala/docs/scaladsl/TransactionsExample.scala) { #transactionalSink }
+: @@ snip [snip](/tests/src/test/scala/docs/4scaladsl/TransactionsExample.scala) { #transactionalSink }
```
Comment (ennru) [inline docs/src/main/paradox/transactions.md]:
```suggestion
: @@ snip [snip](/tests/src/test/scala/docs/scaladsl/TransactionsExample.scala) { #transactionalSink }
```

**Comment by 2m** [comment]
Rebased on top of master and addressed the latest feedback.
